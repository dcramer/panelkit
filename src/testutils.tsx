import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
