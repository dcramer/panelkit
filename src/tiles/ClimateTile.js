import React from "react";
import PropTypes from "prop-types";

import Tile, { TileProps } from "./Tile";
import { toTitleCase } from "../utils";

export default class ClimateTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
  };

  static defaultIcon = "thermomoter";

  getUnit(entityId) {
    if (this.props.unit) return this.props.unit;
    const {
      attributes: { unit_of_measurement },
    } = this.getEntity(entityId);
    return unit_of_measurement || "";
  }

  renderBody() {
    const { attributes } = this.getEntity(this.props.entityId);
    return (
      <div className="tile-climate">
        {attributes.temperature}
        <span className="unit">{this.getUnit(this.props.entityId)}</span>
      </div>
    );
  }

  renderStatus() {
    const { attributes } = this.getEntity(this.props.entityId);
    return (
      <React.Fragment>
        {attributes.current_temperature}
        <span className="unit">{this.getUnit(this.props.entityId)}</span>
      </React.Fragment>
    );
  }

  renderSubtitle() {
    const { attributes } = this.getEntity(this.props.entityId);
    return toTitleCase(attributes.hvac_action);
  }

  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    switch (state) {
      case "heat":
        return "thermometer-high";
      case "cool":
        return "thermometer-low";
      default:
        return super.getDefaultIcon();
    }
  }
}
