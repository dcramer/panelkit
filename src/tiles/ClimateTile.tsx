import React from "react";

import Tile, { TileProps } from "./Tile";
import { toTitleCase } from "../utils";

type ClimateTileProps = TileProps & {
  entityId: string;
  unit?: string;
};

export default class ClimateTile extends Tile<ClimateTileProps> {
  static defaultIcon = "thermomoter";

  getUnit(entityId: string): string {
    if (this.props.unit) return this.props.unit;
    const {
      attributes: { unit_of_measurement },
    } = this.getEntity(entityId);
    return unit_of_measurement || "";
  }

  renderBody() {
    const { attributes } = this.getEntity(this.props.entityId);
    return (
      <div className="tile-climate">
        {attributes.temperature}
        <span className="unit">{this.getUnit(this.props.entityId)}</span>
      </div>
    );
  }

  renderStatus() {
    const { attributes } = this.getEntity(this.props.entityId);
    return (
      <React.Fragment>
        Currently {attributes.current_temperature}
        <span className="unit">{this.getUnit(this.props.entityId)}</span>
      </React.Fragment>
    );
  }

  renderSubtitle() {
    const { attributes } = this.getEntity(this.props.entityId);
    return toTitleCase(attributes.hvac_action);
  }

  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    switch (state) {
      case "heat":
        return "thermometer-high";
      case "cool":
        return "thermometer-low";
      default:
        return super.getDefaultIcon();
    }
  }
}
