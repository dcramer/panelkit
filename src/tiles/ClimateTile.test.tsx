import { renderTile } from "../testutils";
import ClimateTile from "./ClimateTile";

test("renders", async () => {
  const { container, getByText } = renderTile(ClimateTile, {
    entity_id: "climate.test",
    state: "cool",
    attributes: {
      hvac_action: "cooling",
      current_temperature: 75,
      temperature: 70,
      unit_of_measurement: "F",
      friendly_name: "Test Climate",
    },
  });

  expect(getByText(/Test Climate/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
