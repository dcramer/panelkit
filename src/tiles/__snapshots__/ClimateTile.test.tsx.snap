// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-cool false"
  style="cursor: normal;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-status"
    >
      Currently 
      75
      <span
        class="unit"
      >
        F
      </span>
    </div>
    <div
      class="tile-body"
    >
      <div
        class="tile-climate"
      >
        70
        <span
          class="unit"
        >
          F
        </span>
      </div>
    </div>
    <div
      class="tile-title"
    >
      Test Climate
    </div>
    <div
      class="tile-subtitle"
    >
      Cooling
    </div>
  </div>
</div>
`;
